File: benchmarks/FBenchmarks/SpanningTree/Procedural.fs
namespace FBenchmarks.SpanningTree

open System
open System.Collections.Generic
open System.IO

module Procedural =
    open BenchmarkInterface
    type Edge = 
        {
            Start: int;
            End: int;
            Weight: int;
        }
    type Benchmark() =
        // Made by https://www.w3resource.com/csharp-exercises/searching-and-sorting-algorithm/searching-and-sorting-algorithm-exercise-9.php
        let Partition (arr:Edge[]) l r =
            let pivot = arr.[l].Weight;
            let mutable left = l
            let mutable right = r
            let mutable shouldLoop = true
            while shouldLoop do
                while (arr.[left].Weight < pivot) do
                    left <- left + 1

                while (arr.[right].Weight > pivot) do
                    right <- right - 1

                if (left < right) then
                    if arr.[left].Weight = arr.[right].Weight then
                        left <- left + 1
                        right <- right - 1
                    else
                        let temp = arr.[left];
                        arr.[left] <- arr.[right];
                        arr.[right] <- temp;
                else shouldLoop <- false
            right

        let rec QuickSort (arr: Edge []) (left:int) (right:int) =
            if (left < right) then
                let pivot = Partition arr left right

                if (pivot > 1) then
                    QuickSort arr left (pivot - 1)

                if (pivot + 1 < right) then
                    QuickSort arr (pivot + 1) right

        let inputToEdgeArray(input: string []) =
            let c = input.Length
            let mutable arr : Edge array = Array.zeroCreate c
            for i in 0 .. (c-1) do
                let elms = input.[i].Split(',')
                arr.[i] <- { Edge.Start = Int32.Parse(elms.[0]); Edge.End = Int32.Parse(elms.[1]); Edge.Weight = Int32.Parse(elms.[2]) }
            arr

        let vertexGroup : int array = Array.create (6005 + 1) -1
        let rec unionFind (node: int) =
            if vertexGroup .[node] < 0 then
                node
            else 
                vertexGroup .[node] <- unionFind vertexGroup .[node]
                vertexGroup .[node]

        let union (startNode: int) (endNode: int) = 
            let group1Root: int = unionFind startNode
            let group2Root: int = unionFind endNode
            if group1Root = group2Root then
                false
            else
                vertexGroup .[group2Root] <- group1Root
                true

        let computeMinSpanTree (arr: Edge array) = 
            let theMagicNumber: int = (5877-1)
            let result: Edge array = Array.zeroCreate theMagicNumber
            let mutable size = 0
            let mutable totalWeight = 0
            let mutable totalEdges = 0
            let mutable i = 0

            while size < theMagicNumber do
                let currentEdge = arr.[i]
                i <- i + 1
                if (union currentEdge.Start currentEdge.End) then
                    result.[size] <- currentEdge
                    size <- size + 1
                    totalWeight <- totalWeight + currentEdge.Weight
                    totalEdges <- totalEdges + 1

            (totalWeight, totalEdges)
        
        [<DefaultValue>] val mutable input : string []
        interface IBenchmark with
            member this.Preprocess() =
                this.input <- System.IO.File.ReadAllLines("benchmarks/graph.csv")

            member this.Run() =
                let arr = inputToEdgeArray(this.input)
                QuickSort arr 0 (arr.Length - 1)
                let (totalWeight, totalEdges) = computeMinSpanTree arr
                totalWeight+totalEdges

File: benchmarks/CBenchmarks/SpanningTree/Procedural.cs
using BenchmarkInterface;
using System;

namespace CBenchmarks.SpanningTree
{
    public class Procedural : IBenchmark
    {
        public struct Edge
        {
            public int Start;
            public int End;
            public int Weight;
        }

        string[] input;

        public void Preprocess()
        {
            input = System.IO.File.ReadAllLines("benchmarks/graph.csv");
        }

        public int Run()
        {
            //Init the vertex groups
            vertexGroups = new int[6005 + 1];
            for (int i = 0; i < vertexGroups.Length; i++)
            {
                vertexGroups[i] = -1;
            }

            //Do the spanning
            var arr = inputToEdgeArray();
            quick_sort(arr, 0, arr.Length - 1);
            var (weight, edges) = computeMinspanTree(arr);
            return weight+edges;
        }

        Edge[] inputToEdgeArray()
        {
            var c = input.Length;
            Edge[] arr = new Edge[c];
            for (int i = 0; i < c; i++)
            {
                var split = input[i].Split(',');
                arr[i] = new Edge { Start = int.Parse(split[0]), End = int.Parse(split[1]), Weight = int.Parse(split[2]) };
            }
            return arr;
        }

        (int, int) computeMinspanTree(Edge[] arr)
        {
            var magic = 5877 - 1;
            var result = new Edge[magic];
            var size = 0;
            var totalWeight = 0;
            var totalEdges = 0;
            var i = 0;
            while (size < magic)
            {
                var current = arr[i++];
                if (union(current.Start, current.End))
                {
                    result[size] = current;
                    size++;
                    totalWeight += current.Weight;
                    totalEdges++;
                }
            }
            return (totalWeight, totalEdges);
        }

        int[] vertexGroups;
        int unionFind(int node)
        {
            if (vertexGroups[node] < 0)
                return node;
            else
            {
                vertexGroups[node] = unionFind(vertexGroups[node]);
                return vertexGroups[node];
            }
        }

        bool union(int startNode, int endNode)
        {
            var group1Root = unionFind(startNode);
            var group2Root = unionFind(endNode);
            if (group1Root == group2Root)
            {
                return false;
            }
            else
            {
                vertexGroups[group2Root] = group1Root;
                return true;
            }
        }

        // Made by https://www.w3resource.com/csharp-exercises/searching-and-sorting-algorithm/searching-and-sorting-algorithm-exercise-9.php
        void quick_sort(Edge[] arr, int left, int right)
        {
            if (left < right)
            {
                int pivot = partition(arr, left, right);

                if (pivot > 1)
                    quick_sort(arr, left, pivot - 1);

                if (pivot + 1 < right)
                    quick_sort(arr, pivot + 1, right);
            }
        }

        int partition(Edge[] arr, int left, int right)
        {
            int pivot = arr[left].Weight;
            while (true)
            {
                while (arr[left].Weight < pivot)
                    left++;

                while (arr[right].Weight > pivot)
                    right--;

                if (left < right)
                {
                    if (arr[left].Weight == arr[right].Weight)
                    {
                        left++;
                        right--;
                    }
                    else
                    {
                        Edge temp = arr[left];
                        arr[left] = arr[right];
                        arr[right] = temp;
                    }
                }
                else
                    return right;
            }
        }
    }
}